File: IS 643 Quantitative Finance/Signal Processing/kalman.py

x = np.concatenate([x, np.ones([len(stocks['EWA'].values),1])], axis=1)
y = stocks[pair[1]].values.reshape(-1,1)

delta = 0.0001

yhat = [] # Measurement Prediction

e = [] # Measurement prediction error

Q = [] # Measurement prediction error variance

# R(t|t) will be P(t)

P = np.zeros(2,2)

R = np.zeros(2,2)

beta = np.zeros([2,1])

Vw = (delta / (1-delta)) * np.eye(2)

Ve = 0.001

for i in np.arange(0, np.size(y)[0]):
    if i > 0:
        beta = np.concatenate([beta, beta[:,-1].reshape(-1,1)], axis=1) # State prediction
        R = P + Vw # State Covariance Prediction, 3.8

    yhat.append(np.dot(x[i, :], beta[:,i])) # Measurement Prediction, Equation 3.9

    Q.append(x[0,:].dot(R).dot(x[0,:].reshape(-1,1)) + Ve) # Measurement variance prediction, Equation 3.10

    #Observe y(t)
    e.append(y[i] - yhat[i])

    K = R.dot(x[i,:].reshape(-1,1))/Q[i] # Kalman Gain

    beta[:,i] = beta[:,i] + K[:,0]*e[-1] # State update

File: IS 643 Quantitative Finance/Signal Processing/kalman.m
Augment x with ones to accommodate possible offset in the
         % regression
% between y vs x. x=[x ones(size(x))];

delta=0.0001;
% delta=0 allows no change (like traditional
% linear regression).

yhat=NaN(size(y));
% measurement prediction e=NaN(size(y));
% measurement prediction error Q=NaN(size(y));
% measurement prediction error variance

% For clarity, we denote R(t|t) by P(t).
% initialize P and beta.
P=zeros(2);
beta=NaN(2, size(x, 1)); Vw=delta/(1-delta)*diag(ones(2, 1)); Ve=0.001;

% Initialize beta(:, 1) to zero beta(:, 1)=0;
for t=1:length(y)
  if (t > 1)
    beta(:, t)=beta(:, t-1);
    % state prediction.
    % Equation 3.7

    R=P+Vw;
    % state covariance prediction. Equation 3.8
  end
  yhat(t)=x(t, :)*beta(:, t);
  % measurement prediction.
  % Equation 3.9

  Q(t)=x(t, :)*R*x(t, :)’+Ve; % measurement variance % prediction. Equation 3.10

  % Observe y(t)
  e(t)=y(t)-yhat(t); % measurement prediction error
  K=R*x(t, :)’/Q(t); % Kalman gain
  beta(:, t)=beta(:, t)+K*e(t); % State update.
  % Equation 3.11

  P=R-K*x(t, :)*R; % State covariance update. Equation 3.12
End
